### PR TITLE
8320530: has_resolved_ref_index flag not restored after resetting entry

### PR DESCRIPTION
**Pull request by matias9927**
ResolvedMethodEntry::reset_entry() clears the fields in the structure and then restores the constant pool index and the resolved references index if it has one. Currently, the resolved references index is restored without restoring the has_resolved_reference flag used by the structure. 

This patch restored the flag with the resolved references index. Verified with tier 1-5 tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320530](https://bugs.openjdk.org/browse/JDK-8320530): has_resolved_ref_index flag not restored after resetting entry (**Bug** - P3)

### Reviewers
 * [Andrew Dinn](https://openjdk.org/census#adinn) (@adinn - **Reviewer**) ⚠️ Review applies to [0b82817d](https://git.openjdk.org/jdk/pull/16769/files/0b82817ddab43753359efcb7871ca393ae1c36bd)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [0b82817d](https://git.openjdk.org/jdk/pull/16769/files/0b82817ddab43753359efcb7871ca393ae1c36bd)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to [273d82df](https://git.openjdk.org/jdk/pull/16769/files/273d82dfde3f30374da17bf4e8b778ca792f689f)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16769/head:pull/16769` \
`$ git checkout pull/16769`

Update a local copy of the PR: \
`$ git checkout pull/16769` \
`$ git pull https://git.openjdk.org/jdk.git pull/16769/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16769`

View PR using the GUI difftool: \
`$ git pr show -t 16769`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16769.diff">https://git.openjdk.org/jdk/pull/16769.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16769#issuecomment-1823133053)

### CODE REVIEW

**Review by adinn — APPROVED** [review]
Nice catch!

**Review by dholmes-ora — APPROVED** [review]
Change seems fine but what was the effect of not restoring the flag? Does this cause failures or just unnecessary re-resolution, or?

Thanks

**Comment by iklam** [comment]
> Change seems fine but what was the effect of not restoring the flag? Does this cause failures or just unnecessary re-resolution, or?
> 
> Thanks

The code works somehow, but in an unsafe manner. We are reading from `resolved_references_index()` even when the `has_resolved_ref_shift` bit has been (improperly) cleared. Adding the following assert makes it impossible to start jtreg:

```
diff --git a/src/hotspot/share/oops/cpCache.cpp b/src/hotspot/share/oops/cpCache.cpp
index daa094baa7e..760c5268c88 100644
--- a/src/hotspot/share/oops/cpCache.cpp
+++ b/src/hotspot/share/oops/cpCache.cpp
@@ -310,6 +310,9 @@ ResolvedMethodEntry* ConstantPoolCache::set_method_handle(int method_index, cons
   // Store appendix, if any.
   if (has_appendix) {
+    assert(method_entry->has_resolved_ref_index(), "huh");
     const int appendix_index = method_entry->resolved_references_index();
     objArrayOop resolved_references = constant_pool()->resolved_references();
```

I think we should take this as a chance to tighten up the code in resolvedMethodEntry.hpp:

- `resolved_references_index()` should assert that `has_resolved_ref_index()`.
- The following three functions should assert for mutual exclusivity. I.e., you can't call set_klass() and then call  set_resolved_references_index(). Probably the easiest is to add two more bits: `_has_klass_shift` and `_has_table_index_shift`. At entry of these three setters, we should assert that all three klass/table_index/resolved_references_index bits are cleared.

```
  void set_klass(InstanceKlass* klass) {
    _entry_specific._interface_klass = klass;
  }
  void set_resolved_references_index(u2 ref_index) {
    set_flags(1 << has_resolved_ref_shift);
    _entry_specific._resolved_references_index = ref_index;
  }
  void set_table_index(u2 table_index) {
    _entry_specific._table_index = table_index;
  }
```

Also, `has_resolved_ref_index` should be renamed to `has_resolved_reference_index`; otherwise it's difficult to search for all code related to `resolved_refenece_index`.

**Comment by adinn** [comment]
> . . . Probably the easiest is to add two more bits: _has_klass_shift and _has_table_index_shift. . . .
Maybe so, but only in debug builds?

File: src/hotspot/share/oops/cpCache.cpp [hunk 1]
```diff
@@ -309,6 +309,7 @@ ResolvedMethodEntry* ConstantPoolCache::set_method_handle(int method_index, cons
 
   // Store appendix, if any.
   if (has_appendix) {
+    assert(method_entry->has_resolved_references_index(), "sanity");
```
Comment (iklam) [inline src/hotspot/share/oops/cpCache.cpp:312]: The assert is not necessary here, as resolved_references_index() already has the same assert.

**Review by iklam — APPROVED** [review]
LGTM. Just a small nit.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
A few nits with the `DEBUG_ONLY` code.

File: src/hotspot/share/oops/resolvedMethodEntry.hpp [hunk 2]
```diff
@@ -87,7 +91,13 @@ class ResolvedMethodEntry {
       _tos_state(0),
       _flags(0),
       _bytecode1(0),
-      _bytecode2(0) { _entry_specific._interface_klass = nullptr; }
+      _bytecode2(0) {
+        _entry_specific._interface_klass = nullptr;
+#ifdef ASSERT
+        _has_interface_klass = false;
+        _has_table_index = false;
+#endif
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/resolvedMethodEntry.hpp:99]: Indent appears to be 1 instead of 2

**Review by dholmes-ora — APPROVED** [review]
Thanks for the updates.

One indentation nit to fix before integration please.

**Review by adinn — APPROVED** [review]
Still good

**Review by coleenp — CHANGES_REQUESTED** [review]
I have a couple of minor requests before you push.  Hopefully not delaying.

File: src/hotspot/share/oops/resolvedMethodEntry.hpp [hunk 3]
```diff
@@ -77,6 +77,10 @@ class ResolvedMethodEntry {
   u1 _tos_state;                     // TOS state
   u1 _flags;                         // Flags: [00|has_resolved_ref_index|has_local_signature|has_appendix|forced_virtual|final|virtual_final]
   u1 _bytecode1, _bytecode2;         // Resolved invoke codes
+  DEBUG_ONLY(
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/oops/resolvedMethodEntry.hpp]: I see, I said the opposite.  I don't like the #ifdef ASSERT blocks.  3 lines vs 1.

[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/oops/resolvedMethodEntry.hpp]: I think each variable should have DEBUG_ONLY() around it, not grouped though.

**Pull request closed without merging**